src/XPath2/ExtFuncs.cs

```diff
@@ -1026,143 +1026,218 @@ public static object AdjustTimeToTimezone(object arg, object tz)
-                return Math.Floor((double)value);
-            else if (value is Decimal)
-                return Math.Floor((decimal)value);
-            else if (value is Integer)
-                return (Integer)Math.Floor((decimal)(Integer)value);
-            else if (value is Single)
-                return (float)Math.Floor((float)value);
-            else
-                throw new XPath2Exception("XPTY0004", Resources.XPTY0004,
-                    new SequenceType(value.GetType(), XmlTypeCardinality.One),
-                    "xs:float | xs:double | xs:decimal | xs:integer in fn:floor()");
+            }
+
+            switch (value)
+            {
+                case double valueAsDouble:
+                    return Math.Floor(valueAsDouble);
+
+                case decimal valueAsDecimal:
+                    return Math.Floor(valueAsDecimal);
+
+                case Integer valueAsInteger:
+                    return (Integer)Math.Floor((decimal)valueAsInteger);
+
+                case float valueAsFloat:
+                    return (float)Math.Floor(valueAsFloat);
+            }
+
+            throw new XPath2Exception("XPTY0004", Resources.XPTY0004, new SequenceType(value.GetType(), XmlTypeCardinality.One),
+                "xs:float | xs:double | xs:decimal | xs:integer in fn:floor()");
         }
 
         public static object GetRound(object value)
         {
             if (value == Undefined.Value)
+            {
                 return value;
+            }
+
             if (Integer.IsDerivedSubtype(value))
+            {
                 value = Integer.ToInteger(value);
-            if (value is Double)
-                return Math.Round((double)value);
-            else if (value is Decimal)
-                return Math.Round((decimal)value);
-            else if (value is Integer)
-                return (Integer)Math.Round((decimal)(Integer)value);
-            else if (value is Single)
-                return (float)Math.Round((float)value);
-            else
-                throw new XPath2Exception("XPTY0004", Resources.XPTY0004,
-                    new SequenceType(value.GetType(), XmlTypeCardinality.One),
-                    "xs:float | xs:double | xs:decimal | xs:integer in fn:round()");
+            }
+
+            switch (value)
+            {
+                case double valueAsDouble:
+                    double fractionAsDouble = Math.Abs(valueAsDouble - Math.Floor(valueAsDouble));
+                    return fractionAsDouble >= 0.5 ? Math.Ceiling(valueAsDouble) : Math.Floor(valueAsDouble);
+
+                case decimal valueAsDecimal:
+                    decimal fractionAsDecimal = Math.Abs(valueAsDecimal - Math.Floor(valueAsDecimal));
+                    return fractionAsDecimal >= 0.5m ? Math.Ceiling(valueAsDecimal) : Math.Floor(valueAsDecimal);
+
+                case Integer valueAsInteger:
+                    decimal valueCastToDecimal = (decimal)valueAsInteger;

```

Comment:
Maybe I'm missing something but if the value is already an integer we don't need to round it right? Could just return valueAsInteger?